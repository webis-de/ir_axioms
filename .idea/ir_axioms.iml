<?xml version="1.0" encoding="UTF-8"?>
<module type="PYTHON_MODULE" version="4">
  <component name="NewModuleRootManager">
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/venv" />
      <excludeFolder url="file://$MODULE_DIR$/build" />
      <excludeFolder url="file://$MODULE_DIR$/dist" />
      <excludeFolder url="file://$MODULE_DIR$/venv-base/lib" />
      <excludeFolder url="file://$MODULE_DIR$/.pytest_cache" />
      <excludeFolder url="file://$MODULE_DIR$/ir_axioms.egg-info" />
      <excludeFolder url="file://$MODULE_DIR$/venv-base" />
      <excludeFolder url="file://$MODULE_DIR$/tmp" />
      <excludeFolder url="file://$MODULE_DIR$/venv-pyserini" />
      <excludeFolder url="file://$MODULE_DIR$/venv-pyterrier" />
      <excludeFolder url="file://$MODULE_DIR$/data" />
      <excludeFolder url="file://$MODULE_DIR$/venv-test" />
      <excludeFolder url="file://$MODULE_DIR$/.idea/dataSources" />
      <excludeFolder url="file://$MODULE_DIR$/examples/cache" />
      <excludeFolder url="file://$MODULE_DIR$/examples/trec-system-runs" />
      <excludeFolder url="file://$MODULE_DIR$/examples/pyterrier" />
      <excludeFolder url="file://$MODULE_DIR$/examples/data" />
      <excludeFolder url="file://$MODULE_DIR$/examples/results" />
      <excludeFolder url="file://$MODULE_DIR$/venv-3.7" />
      <excludeFolder url="file://$MODULE_DIR$/examples/.ipynb_checkpoints" />
      <excludeFolder url="file://$MODULE_DIR$/examples/msmarco-index" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
  <component name="PackageRequirementsSettings">
    <option name="requirementsPath" value="$MODULE_DIR$/setup.cfg" />
  </component>
</module>