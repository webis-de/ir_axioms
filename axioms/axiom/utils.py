from typing import TypeVar, Union

T = TypeVar("T")


def strictly_greater(x: T, y: T) -> float:
    if x > y:
        return 1
    elif y > x:
        return -1
    return 0


def strictly_less(x: T, y: T) -> float:
    if y > x:
        return 1
    elif x > y:
        return -1
    return 0


def approximately_equal(
        *items: Union[int, float],
        margin_fraction: float = 0.1
) -> bool:
    """
    True if all numeric args are
    within (100 * margin_fraction)% of the largest.
    """
    if len(items) == 0:
        return True

    abs_max = max(items, key=lambda item: abs(item))
    if abs_max == 0:
        # All values are 0.
        return True

    boundaries = (
        abs_max * (1 + margin_fraction),
        abs_max * (1 - margin_fraction),
    )
    boundary_min = min(boundaries)
    boundary_max = max(boundaries)

    return all(boundary_min < item < boundary_max for item in items)
