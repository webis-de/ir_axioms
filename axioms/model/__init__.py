# Re-export from sub-modules.

from axioms.model.base import (  # noqa: F401
    Input,
    Output,
    Preference,
    PreferenceMatrix,
    Mask,
    MaskMatrix,
)

from axioms.model.retrieval import (  # noqa: F401
    Query,
    TextQuery,
    Document,
    TextDocument,
    ScoredDocument,
    ScoredTextDocument,
    RankedDocument,
    JudgedDocument,
    IndexContext,
)

from axioms.model.generation import (  # noqa: F401
    GenerationInput,
    GenerationOutput,
    Aspect,
    Aspects,
)
