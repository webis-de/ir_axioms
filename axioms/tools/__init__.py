from injector import Module, Binder

# Re-export from sub-modules.

from axioms.tools.aspects import (  # noqa: F401
    AspectExtraction,
    SpacyNounChunksAspectExtraction,
    AspectsModule,
)

from axioms.tools.contents import (  # noqa: F401
    TextContents,
    DocumentQueryTextContents,
    IrdsDocumentTextContents,
    IrdsQueryTextContents,
    AnseriniDocumentTextContents,
    TerrierDocumentTextContents,
    HasText,
    SimpleTextContents,
    ContentsModule,
)

from axioms.tools.index_statistics import (  # noqa: F401
    IndexStatistics,
    AnseriniIndexStatistics,
    TerrierIndexStatistics,
)

from axioms.tools.pivot import (  # noqa: F401
    PivotSelection,
    RandomPivotSelection,
    FirstPivotSelection,
    LastPivotSelection,
    MiddlePivotSelection,
    PivotModule,
)

from axioms.tools.similarity import (  # noqa: F401
    TermSimilarity,
    SentenceSimilarity,
    FastTextTermSimilarity,
    WordNetSynonymSetTermSimilarity,
    SentenceTransformersSentenceSimilarity,
    WordNetSynonymSetTermSimilarity,
    SimilarityModule,
)

from axioms.tools.text_statistics import (  # noqa: F401
    TextStatistics,
    DocumentQueryTextStatistics,
    AnseriniTextStatistics,
    TerrierTextStatistics,
    SimpleTextStatistics,
    TextStatisticsModule,
)

from axioms.tools.tokenizer import (  # noqa: F401
    TermTokenizer,
    SentenceTokenizer,
    NltkTermTokenizer,
    NltkSentenceTokenizer,
    AnseriniTermTokenizer,
    TerrierTermTokenizer,
    TokenizerModule,
)


class ToolsModule(Module):
    def configure(self, binder: Binder) -> None:
        binder.install(AspectsModule)
        binder.install(ContentsModule)
        binder.install(PivotModule)
        binder.install(SimilarityModule)
        binder.install(TokenizerModule)
        # Need to be loaded after the tokenizer module because it needs it.
        binder.install(TextStatisticsModule)
