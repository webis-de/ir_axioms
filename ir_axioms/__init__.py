from logging import basicConfig, Logger, getLogger

from class_registry import SortedClassRegistry
from importlib_metadata import version

__version__ = version("ir_axioms")

basicConfig()
logger: Logger = getLogger("ir_axioms")

registry = SortedClassRegistry(
    unique=True,
    attr_name="name",
    sort_key="name",
)
