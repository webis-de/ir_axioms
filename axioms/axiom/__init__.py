# Re-export from sub-modules.

from axioms.axiom.generation import (  # noqa: F401
    AspectCoverageAxiom,
)

from axioms.axiom.retrieval import (  # noqa: F401
    ArgumentativeUnitsCountAxiom,
    ArgUC,
    QueryTermOccurrenceInArgumentativeUnitsAxiom,
    QTArg,
    QueryTermPositionInArgumentativeUnitsAxiom,
    QTPArg,
    AverageSentenceLengthAxiom,
    aSLDoc,
    aSL,
    Lnc1Axiom,
    LNC1,
    TfLncAxiom,
    TF_LNC,
    Lb1Axiom,
    LB1,
    Prox1Axiom,
    PROX1,
    Prox2Axiom,
    PROX2,
    Prox3Axiom,
    PROX3,
    Prox4Axiom,
    PROX4,
    Prox5Axiom,
    PROX5,
    RegAxiom,
    REG,
    AntiRegAxiom,
    ANTI_REG,
    AspectRegAxiom,
    ASPECT_REG,
    AndAxiom,
    AND,
    LEN_AND,
    ModifiedAndAxiom,
    M_AND,
    LEN_M_AND,
    DivAxiom,
    DIV,
    LEN_DIV,
    OriginalAxiom,
    ORIG,
    OracleAxiom,
    ORACLE,
    Tfc1Axiom,
    TFC1,
    Tfc3Axiom,
    TFC3,
    ModifiedTdcAxiom,
    M_TDC,
    LEN_M_TDC,
    Stmc1Axiom,
    STMC1,
    Stmc2Axiom,
    STMC2,
    TrecOracleAxiom,
)


from axioms.axiom.arithmetic import (  # noqa: F401
    UniformAxiom,
    SumAxiom,
    ProductAxiom,
    MultiplicativeInverseAxiom,
    ConjunctionAxiom,
    VoteAxiom,
    MajorityVoteAxiom,
    NormalizedAxiom,
)

from axioms.axiom.base import (  # noqa: F401
    Axiom,
)

from axioms.axiom.cache import (  # noqa: F401
    CachedAxiom,
)

from axioms.axiom.estimator import (  # noqa: F401
    EstimatorAxiom,
    ScikitLearnEstimator,
    ScikitLearnEstimatorAxiom,
)

from axioms.axiom.parallel import (  # noqa: F401
    ParllelAxiom,
)

from axioms.axiom.precondition import (  # noqa: F401
    PreconditionAxiom,
)

from axioms.axiom.simple import (  # noqa: F401
    NopAxiom,
    NOP,
    RandomAxiom,
    RANDOM,
    GreaterThanAxiom,
    GT,
    LessThanAxiom,
    LT,
)
