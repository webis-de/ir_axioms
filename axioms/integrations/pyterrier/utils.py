from axioms.utils.libraries import is_pyterrier_installed

if is_pyterrier_installed():
    from dataclasses import dataclass
    from pathlib import Path
    from typing import Set, Optional, Sequence, Callable, Any, Union, cast

    from injector import singleton
    from ir_datasets import Dataset
    from pandas import DataFrame, Series
    from pyterrier import Transformer
    from pyterrier.datasets import IRDSDataset
    from pyterrier.java import (
        required as pt_java_required,
        autoclass as pt_java_autoclass,
    )

    from axioms.dependency_injection import injector
    from axioms.model import (
        RankedScoredDocument,
        RankedScoredTextDocument,
        JudgedRankedScoredTextDocument,
        JudgedRankedScoredDocument,
        Query,
        Document,
    )
    from axioms.tools import (
        TextContents,
        SimpleTextContents,
        IrdsQueryTextContents,
        IrdsDocumentTextContents,
        TerrierDocumentTextContents,
        TextStatistics,
        TerrierTextStatistics,
        TermTokenizer,
        TerrierTermTokenizer,
        IndexStatistics,
        TerrierIndexStatistics,
    )

    @pt_java_required
    def autoclass(*args, **kwargs) -> Any:
        return pt_java_autoclass(*args, **kwargs)

    Index = autoclass("org.terrier.structures.Index")
    IndexRef = autoclass("org.terrier.querying.IndexRef")
    Tokeniser = autoclass("org.terrier.indexing.tokenisation.Tokeniser")
    EnglishTokeniser = autoclass("org.terrier.indexing.tokenisation.EnglishTokeniser")

    def inject_pyterrier(
        index_location: Union[Index, IndexRef, Path, str],  # type: ignore
        text_field: Optional[str] = "text",
        tokeniser: Tokeniser = EnglishTokeniser(),  # type: ignore
        dataset: Optional[Union[Dataset, str, IRDSDataset]] = None,
    ) -> None:
        injector.binder.bind(
            interface=TextStatistics,
            to=TerrierTextStatistics(index_location=index_location),
            scope=singleton,
        )
        injector.binder.bind(
            interface=TermTokenizer,
            to=TerrierTermTokenizer(tokeniser=tokeniser),
            scope=singleton,
        )
        injector.binder.bind(
            interface=IndexStatistics,
            to=TerrierIndexStatistics(index_location=index_location),
            scope=singleton,
        )

        if text_field is not None:
            injector.binder.bind(
                interface=TextContents[Document],
                to=TerrierDocumentTextContents(
                    simple_text_contents=injector.get(SimpleTextContents),
                    index_location=index_location,
                    text_field=text_field,
                ),
                scope=singleton,
            )

        if dataset is not None:
            injector.binder.bind(
                interface=TextContents[Query],
                to=IrdsQueryTextContents(
                    simple_text_contents=injector.get(SimpleTextContents),
                    dataset=(
                        dataset
                        if not isinstance(dataset, IRDSDataset)
                        else cast(Dataset, dataset.irds_ref)
                    ),
                ),
                scope=singleton,
            )
            injector.binder.bind(
                interface=TextContents[Document],
                to=IrdsDocumentTextContents(
                    simple_text_contents=injector.get(SimpleTextContents),
                    dataset=(
                        dataset
                        if not isinstance(dataset, IRDSDataset)
                        else cast(Dataset, dataset.irds_ref)
                    ),
                ),
                scope=singleton,
            )

    def require_columns(
        ranking: DataFrame,
        expected_columns: Set[str],
    ) -> None:
        columns: Set[str] = set(ranking.columns)
        missing_columns: Set[str] = expected_columns - columns
        if len(missing_columns) > 0:
            raise ValueError(
                f"Expected columns "
                f"{', '.join(expected_columns)} but got columns "
                f"{', '.join(columns)} (missing columns "
                f"{', '.join(missing_columns)})."
            )

    def load_documents(
        ranking: DataFrame,
        text_column: Optional[str] = "text",
    ) -> Sequence[RankedScoredDocument]:
        require_columns(ranking, {"docno", "rank", "score"})

        parser: Callable[[Series], RankedScoredDocument]

        if "label" in ranking.columns:
            if text_column is not None:

                def parser(row: Series) -> RankedScoredDocument:
                    return JudgedRankedScoredTextDocument(
                        id=str(row["docno"]),
                        text=str(row[text_column]),
                        score=float(row["score"]),
                        rank=int(row["rank"]),
                        relevance=float(row["label"]),
                    )

            else:

                def parser(row: Series) -> RankedScoredDocument:
                    return JudgedRankedScoredDocument(
                        id=str(row["docno"]),
                        score=float(row["score"]),
                        rank=int(row["rank"]),
                        relevance=float(row["label"]),
                    )

        else:
            if text_column is not None:

                def parser(row: Series) -> RankedScoredDocument:
                    return RankedScoredTextDocument(
                        id=str(row["docno"]),
                        text=str(row[text_column]),
                        score=float(row["score"]),
                        rank=int(row["rank"]),
                    )

            else:

                def parser(row: Series) -> RankedScoredDocument:
                    return RankedScoredDocument(
                        id=str(row["docno"]),
                        score=float(row["score"]),
                        rank=int(row["rank"]),
                    )

        return [parser(row) for _, row in ranking.iterrows()]

    def load_queries(ranking: DataFrame) -> Sequence[Query]:
        require_columns(ranking, {"query"})
        return [Query(row["query"]) for _, row in ranking.iterrows()]

    @dataclass(frozen=True)
    class FilterTopicsTransformer(Transformer):
        """
        Retain only queries that are contained in the topics.
        """

        topics: DataFrame

        def transform(self, ranking: DataFrame) -> DataFrame:
            return ranking[ranking["qid"].isin(self.topics["qid"])]

    @dataclass(frozen=True)
    class FilterQrelsTransformer(Transformer):
        """
        Retain only query-document pairs that are contained in the qrels.
        """

        qrels: DataFrame

        def transform(self, ranking: DataFrame) -> DataFrame:
            return ranking[
                ranking["qid"].isin(self.qrels["qid"])
                & ranking["docno"].isin(self.qrels["docno"])
            ]

    @dataclass(frozen=True)
    class JoinQrelsTransformer(Transformer):
        """
        Join query-document pairs with their relevance labels.
        """

        qrels: DataFrame

        def transform(self, ranking: DataFrame) -> DataFrame:
            qrels = self.qrels
            require_columns(qrels, {"qid", "docno", "label"})
            return ranking.merge(self.qrels, on=["qid", "docno"], how="left")

    @dataclass(frozen=True)
    class AddNameTransformer(Transformer):
        """
        Add a colum containing the system's name.
        """

        system_name: str

        def transform(self, res: DataFrame) -> DataFrame:
            res["name"] = self.system_name
            return res

else:
    inject_pyterrier = NotImplemented
    require_columns = NotImplemented
    load_documents = NotImplemented
    load_queries = NotImplemented
    FilterTopicsTransformer = NotImplemented  # type: ignore
    FilterQrelsTransformer = NotImplemented  # type: ignore
    JoinQrelsTransformer = NotImplemented  # type: ignore
    AddNameTransformer = NotImplemented  # type: ignore
