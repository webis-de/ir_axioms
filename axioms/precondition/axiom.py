from dataclasses import dataclass
from typing import Generic, Literal, Sequence, TYPE_CHECKING

from numpy import sign

from axioms.model import (
    Input,
    Output,
    Mask,
    MaskMatrix,
)
from axioms.precondition.base import Precondition

if TYPE_CHECKING:
    from axioms.axiom.base import Axiom


@dataclass(frozen=True, kw_only=True)
class AxiomPrecondition(Precondition[Input, Output], Generic[Input, Output]):
    axiom: "Axiom[Input, Output]"
    expected_sign: Literal[1, 0, -1] = 0
    strip_preconditions: bool = True

    def precondition(
        self,
        input: Input,
        output1: Output,
        output2: Output,
    ) -> Mask:
        preference = self.axiom.preference(input, output1, output2)
        return sign(preference) == self.expected_sign

    def preconditions(
        self,
        input: Input,
        outputs: Sequence[Output],
    ) -> MaskMatrix:
        preferences = self.axiom.preferences(input, outputs)
        return sign(preferences) == self.expected_sign
