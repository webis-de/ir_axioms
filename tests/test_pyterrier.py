from typing import Any, TYPE_CHECKING

from approvaltests import verify
from pytest import skip, fixture

from ir_axioms.axiom import Axiom
from ir_axioms.model import Document
from ir_axioms.axiom.utils import strictly_greater
from ir_axioms.tools import MiddlePivotSelection
from ir_axioms.utils.libraries import is_pyterrier_installed


class _DocumentIdAxiom(Axiom[Any, Document]):
    def preference(self, input: Any, output1: Document, output2: Document) -> float:
        return strictly_greater(output1.id, output2.id)


_DOC_ID = _DocumentIdAxiom()


def test_kwiksort_empty_input() -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    from pandas import DataFrame
    from ir_axioms.integrations import KwikSortReranker

    df_in = DataFrame(columns=["qid", "docno"])

    kwiksort = KwikSortReranker(
        axiom=_DOC_ID,
        pivot_selection=MiddlePivotSelection(),
    )

    df_out = kwiksort.transform(df_in)
    verify(df_out.to_csv(path_or_buf=None, index=False, header=True))


def test_kwiksort_rerank() -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    from pandas import DataFrame
    from ir_axioms.integrations import KwikSortReranker

    df_in = DataFrame(
        [
            {"qid": "q1", "docno": "doc1"},
            {"qid": "q1", "docno": "doc3"},
            {"qid": "q1", "docno": "doc2"},
            {"qid": "q2", "docno": "doc5"},
            {"qid": "q2", "docno": "doc6"},
            {"qid": "q2", "docno": "doc4"},
        ]
    )

    kwiksort = KwikSortReranker(
        axiom=_DOC_ID,
        pivot_selection=MiddlePivotSelection(),
    )

    df_out = kwiksort.transform(df_in)
    verify(df_out.to_csv(path_or_buf=None, index=False, header=True))


def test_axiomatic_preferences_empty_input() -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    from pandas import DataFrame
    from ir_axioms.integrations import AxiomaticPreferences

    df_in = DataFrame(columns=["qid", "docno"])

    axiomatic_preferences = AxiomaticPreferences(
        axioms=[_DOC_ID],
        axiom_names=["_DOC_ID"],
        text_field="text",
    )

    df_out = axiomatic_preferences.transform(df_in)
    verify(df_out.to_csv(path_or_buf=None, index=False, header=True))


def test_axiomatic_preferences_compute() -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    from pandas import DataFrame
    from ir_axioms.integrations import AxiomaticPreferences

    df_in = DataFrame(
        [
            {"qid": "q1", "docno": "doc1"},
            {"qid": "q1", "docno": "doc3"},
            {"qid": "q1", "docno": "doc2"},
            {"qid": "q2", "docno": "doc5"},
            {"qid": "q2", "docno": "doc6"},
            {"qid": "q2", "docno": "doc4"},
        ]
    )

    axiomatic_preferences = AxiomaticPreferences(
        axioms=[_DOC_ID],
        axiom_names=["_DOC_ID"],
        text_field="text",
    )

    df_out = axiomatic_preferences.transform(df_in)
    verify(df_out.to_csv(path_or_buf=None, index=False, header=True))


def test_aggregated_axiomatic_preferences_empty_input() -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    from pandas import DataFrame
    from ir_axioms.integrations import AggregatedAxiomaticPreferences

    df_in = DataFrame(columns=["qid", "docno"])

    aggregated_axiomatic_preferences = AggregatedAxiomaticPreferences(
        axioms=[_DOC_ID],
        text_field="text",
    )

    df_out = aggregated_axiomatic_preferences.transform(df_in)
    verify(df_out.to_csv(path_or_buf=None, index=False, header=True))


def test_aggregated_axiomatic_preferences_compute() -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    from pandas import DataFrame
    from ir_axioms.integrations import AggregatedAxiomaticPreferences

    df_in = DataFrame(
        [
            {"qid": "q1", "docno": "doc1"},
            {"qid": "q1", "docno": "doc3"},
            {"qid": "q1", "docno": "doc2"},
            {"qid": "q2", "docno": "doc5"},
            {"qid": "q2", "docno": "doc6"},
            {"qid": "q2", "docno": "doc4"},
        ]
    )

    aggregated_axiomatic_preferences = AggregatedAxiomaticPreferences(
        axioms=[_DOC_ID],
        text_field="text",
    )

    df_out = aggregated_axiomatic_preferences.transform(df_in)
    verify(df_out.to_csv(path_or_buf=None, index=False, header=True))


# TODO: Estimator

if TYPE_CHECKING:
    from ir_axioms.integrations import AxiomaticExperiment
else:
    AxiomaticExperiment = Any


@fixture
def axiomatic_experiment() -> AxiomaticExperiment:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    from pandas import DataFrame
    from pyterrier import Transformer

    from ir_axioms.integrations import AxiomaticExperiment

    return AxiomaticExperiment(
        retrieval_systems=[
            Transformer.from_df(
                DataFrame(
                    [
                        {"qid": "q1", "docno": "doc2", "score": 0.5, "rank": 0},
                        {"qid": "q1", "docno": "doc1", "score": 0.3, "rank": 1},
                        {"qid": "q2", "docno": "doc1", "score": 0.4, "rank": 0},
                        {"qid": "q2", "docno": "doc2", "score": 0.2, "rank": 1},
                    ]
                )
            )
        ],
        names=["test_system"],
        topics=DataFrame(
            [
                {"qid": "q1"},
                {"qid": "q2"},
            ]
        ),
        qrels=DataFrame(
            [
                {"qid": "q1", "docno": "doc1", "label": 1},
                {"qid": "q1", "docno": "doc2", "label": 0},
                {"qid": "q2", "docno": "doc1", "label": 1},
                {"qid": "q2", "docno": "doc2", "label": 0},
            ]
        ),
        axioms=[_DOC_ID],
        axiom_names=["_DOC_ID"],
        text_field="text",
    )


def test_axiomatic_experiment_preferences(
    axiomatic_experiment: AxiomaticExperiment,
) -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    verify(
        axiomatic_experiment.preferences.to_csv(
            path_or_buf=None, index=False, header=True
        )
    )


def test_axiomatic_experiment_preference_distribution(
    axiomatic_experiment: AxiomaticExperiment,
) -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    verify(
        axiomatic_experiment.preference_distribution.to_csv(
            path_or_buf=None, index=False, header=True
        )
    )


def test_axiomatic_experiment_preference_consistency(
    axiomatic_experiment: AxiomaticExperiment,
) -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    verify(
        axiomatic_experiment.preference_consistency.to_csv(
            path_or_buf=None, index=False, header=True
        )
    )


def test_axiomatic_experiment_inconsistent_pairs(
    axiomatic_experiment: AxiomaticExperiment,
) -> None:
    if not is_pyterrier_installed():
        skip("PyTerrier is not installed.")

    verify(
        axiomatic_experiment.inconsistent_pairs.to_csv(
            path_or_buf=None, index=False, header=True
        )
    )
