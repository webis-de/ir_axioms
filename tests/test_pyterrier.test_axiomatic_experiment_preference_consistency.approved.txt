axiom,ORIG_consistency,ORACLE_consistency
ORIG,1.0,0.5
ORACLE,0.5,1.0
_DOC_ID,0.5,0.0
